app/gui/src/project-view/components/GraphEditor/widgets/WidgetSelection/SelectionSubmenu.vue

```diff
@@ -0,0 +1,139 @@
+<script setup lang="ts">
+import SizeTransition from '@/components/SizeTransition.vue'
+import DropdownWidget, { DropdownEntry } from '@/components/widgets/DropdownWidget.vue'
+import { unrefElement } from '@/composables/events'
+import { targetIsOutside } from '@/util/autoBlur'
+import { computed, ComputedRef, ref, useTemplateRef, watch } from 'vue'
+import { submenuDropdownStyles } from './styles'
+import { Entry, ExpressionTag, isEntry, NestedChoiceTag } from './tags'
+
+const props = defineProps<{
+  rootElement: HTMLElement | undefined
+  floatReference: HTMLElement | undefined
+  show: boolean
+  entries: Entry[]
+  selectedExpressions: Set<string>
+  topLevel?: boolean
+}>()
+
+const emit = defineEmits<{
+  clickedEntry: [Entry, boolean]
+}>()
+
+const dropdownElement = useTemplateRef('dropdownElement')
+const floatReference = computed(() => props.floatReference)
+const rootElement = computed(() => props.rootElement)
+
+const { floatingStyles } = submenuDropdownStyles(
+  floatReference,
+  dropdownElement,
+  props.topLevel,
+  rootElement,
+)
+
+/** Referring to the type of the component in the current file is hard, so we define a helper type. */
+interface SubmenuComponent {
+  isTargetOutside: (event: Event) => boolean
+}
+function isSubmenuComponent(component: unknown): component is SubmenuComponent {
+  return (
+    component != null &&
+    typeof component === 'object' &&
+    'isTargetOutside' in component &&
+    typeof component.isTargetOutside === 'function'
+  )
+}
+
+/** Check if the event target is outside the current submenu and any of its descendants. */
+function isTargetOutside(event: Event) {
+  const isOutsideCurrent = targetIsOutside(event, unrefElement(dropdownElement))
+  const isOutsideSubmenu =
+    isSubmenuComponent(submenuRef.value) ? submenuRef.value.isTargetOutside(event) : true
+  return isOutsideCurrent && isOutsideSubmenu
+}
+
+defineExpose({
+  isTargetOutside,
+})
+
+export interface Submenu {
+  entries: ComputedRef<Entry[]>
+  relativeTo: HTMLElement
+}
+
+const submenu = ref<Submenu | null>(null)
+const submenuEntries = computed(() => submenu.value?.entries ?? [])
+const submenuRef = useTemplateRef('submenuRef')
```

Comment:
I propose a bit of reordering here, for better reading: types at the top (after props maybe), then refs, then rest and expose at end (the last is not a rule per se, but a place I usually look for exposes).

Reply:
Did some reordering, trying to keep related parts together.